### PR TITLE
[stdlib] Documentation improvements

### PR DESCRIPTION
**Pull request by natecook1000**
- Revise Equatable and Hashable for synthesized requirements
- Complete Strideable and stride(from:...:by:) documentation
- Revise DoubleWidth type docs
- Add complexity notes for Set.index(of:) and .contains(_:)
- Fix typos in Set.formUnion docs
- Add missing axioms for SetAlgebra (SR-6319)
- Improve guidance for description and debugDescription
- Add note about the result of passing duplicate keys to
  Dictionary(uniqueKeysWithValues:)
- Fix typo in BinaryInteger docs
- Update Substring docs with better conversion example
- Improve docs for withMemoryRebound and isKnownUniquelyReferenced

### CODE REVIEW

File: stdlib/public/core/Equatable.swift [hunk 1]
```diff
@@ -48,17 +48,28 @@
 /// `Comparable` protocols, which allow more uses of your custom type, such as
 /// constructing sets or sorting the elements of a collection.
 ///
```
Comment (natecook1000) [inline stdlib/public/core/Equatable.swift]: @allevato Would you mind taking a look at this section and the changes to `Hashable.swift`? Thanks!

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Review by airspeedswift — APPROVED** [review]
LGTM

File: stdlib/public/core/Equatable.swift (hunk 1 above)
Comment (allevato) [reply stdlib/public/core/Equatable.swift]: Looks great! Thanks for updating all of this.

**Pull request merged**